nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py

```diff
@@ -107,6 +110,105 @@ def write_entry(profile, generation, machine_id):
             f.write("machine-id %s\n" % machine_id)
     os.rename(tmp_path, entry_file)
 
+def sb_efi_file_name_relative(profile, generation):
+    if profile:
+        return "efi/nixos/nixos-%s-generation-%d.efi" % (profile, generation)
+    else:
+        return "efi/nixos/nixos-generation-%d.efi" % (generation)
+
+def make_signed_efi(profile, generation, efi_file):
+    with tempfile.TemporaryDirectory() as tmpdir:
+        append_initrd_secrets = profile_path(profile, generation, "append-initrd-secrets")
+        if os.path.exists(append_initrd_secrets):
+            initrd = f"{tmpdir}/initrd"
+            shutil.copyfile(
+                profile_path(profile, generation, "initrd"),
+                initrd
+            )
+
+            subprocess.check_call([append_initrd_secrets, "@efiSysMountPoint@%s" % (initrd)])
+        else:
+            initrd = profile_path(profile, generation, "initrd"),
+
+        generation_dir = os.readlink(system_dir(profile, generation))
+        tmp_path = "%s/efistub" % (tmpdir)
+
+        kernel_params = "systemConfig=%s init=%s/init " % (generation_dir, generation_dir)
+        with open("%s/kernel-params" % (generation_dir)) as params_file:
+            kernel_params = kernel_params + params_file.read()
+        kernel_param_file = "%s/kernel_params" % tmpdir
+        with open(kernel_param_file, 'w') as f:
+            f.write(kernel_params)
+
+        subprocess.check_call([
+            "@binutils@/bin/objcopy",
+            "--add-section", ".osrel={}/etc/os-release".format(generation_dir), "--change-section-vma", ".osrel=0x20000",
+            "--add-section", ".cmdline={}".format(kernel_param_file),           "--change-section-vma", ".cmdline=0x30000",
+            "--add-section", ".linux={}/kernel".format(generation_dir),         "--change-section-vma", ".linux=0x40000",
+            "--add-section", ".initrd={}".format(initrd),                       "--change-section-vma", ".initrd=0x3000000",
+            "{}/sw/lib/systemd/boot/efi/linuxx64.efi.stub".format(generation_dir),
+            tmp_path
+        ])
+        sign_path(tmp_path, efi_file)
+
+
+def write_secureboot_entry(profile, generation, machine_id):
+    if profile:
+        entry_file = "@efiSysMountPoint@/loader/entries/nixos-%s-generation-%d.conf" % (profile, generation)
+    else:
+        entry_file = "@efiSysMountPoint@/loader/entries/nixos-generation-%d.conf" % (generation)
+    efi_file_relative = sb_efi_file_name_relative(profile, generation)
+    efi_file = "@efiSysMountPoint@/%s" % (efi_file_relative)
+
+    try:
+        sbverify(efi_file)
```

Comment:
This is not a valid/safe statement, because:

```
$ sudo rm /nix/var/nix/profiles/system-37-link # latest
$ sudo nixos-rebuild boot
$ sudo rm /nix/var/nix/profiles/system-37-link # latest
$ sudo nixos-rebuild boot --upgrade
$ nix-collect-garbage
```

will make boot option 37 unbootable. Maybe storing it by the hash part of the generation's store path root would work better.